compiler/lib/effects.ml

```diff
@@ -0,0 +1,415 @@
+  free_pc
+
+let closure_of_pc ~st pc =
+  try Addr.Map.find pc st.jc.closure_of_jump with Not_found -> assert false
+
+let allocate_closure ~st ~params ~body ~branch =
+  let block = { params = []; body; branch } in
+  let pc = add_block st block in
+  let name = Var.fresh () in
+  [ Let (name, Closure (params, (pc, []))) ], name
+
+let cps_branch ~st (pc, args) =
+  let ret = Var.fresh () in
+  [ Let (ret, Apply { f = closure_of_pc ~st pc; args; exact = true }) ], Return ret
+
+let cps_jump_cont ~st cont =
+  let call_block =
+    let body, branch = cps_branch ~st cont in
+    add_block st { params = []; body; branch }
+  in
+  call_block, []
+
+let cps_last ~st (last : last) ~k : instr list * last =
+  match last with
+  | Return x ->
+      let ret = Var.fresh () in
+      [ Let (ret, Apply { f = k; args = [ x ]; exact = true }) ], Return ret
+  | Raise (x, _) ->
+      let ret = Var.fresh () in
+      let exn_handler = Var.fresh_n "raise" in
+      ( [ Let (exn_handler, Prim (Extern "caml_pop_trap", []))
+        ; Let (ret, Apply { f = exn_handler; args = [ x ]; exact = true })
+        ]
+      , Return ret )
+  | Stop -> [], Stop
+  | Branch cont -> cps_branch ~st cont
+  | Cond (x, cont1, cont2) ->
+      assert (not (Poly.equal cont1 cont2));
+      [], Cond (x, cps_jump_cont ~st cont1, cps_jump_cont ~st cont2)
+  | Switch (x, c1, c2) ->
+      (* To avoid code duplication during JavaScript generation, we need
+         to create a single block per continuation *)
+      let cps_jump_cont = Fun.memoize (cps_jump_cont ~st) in
+      [], Switch (x, Array.map c1 ~f:cps_jump_cont, Array.map c2 ~f:cps_jump_cont)
+  | Pushtrap ((pc, args), x, handler_cont, _) ->
+      let constr_handler, exn_handler =
+        (* Construct handler closure *)
+        allocate_closure ~st ~params:[ x ] ~body:[] ~branch:(Branch handler_cont)
+      in
+      let ret = Var.fresh () in
+      ( constr_handler
+        @ [ Let (Var.fresh (), Prim (Extern "caml_push_trap", [ Pv exn_handler ]))
+          ; Let (ret, Apply { f = closure_of_pc ~st pc; args; exact = true })
+          ]
+      , Return ret )
+  | Poptrap (pc, args) ->
+      let ret = Var.fresh () in
+      let exn_handler = Var.fresh () in
+      ( [ Let (exn_handler, Prim (Extern "caml_pop_trap", []))
+        ; Let (ret, Apply { f = closure_of_pc ~st pc; args; exact = true })
+        ]
+      , Return ret )
+
+let cps_instr ~st (instr : instr) : instr =
+  match instr with
+  | Let (x, Closure (params, (pc, args))) ->
+      Let (x, Closure (params @ [ st.closure_continuation pc ], (pc, args)))
+  | Let (x, Prim (Extern "caml_alloc_dummy_function", [ size; arity ])) -> (
+      match arity with
+      | Pc (Int a) ->
+          Let
+            ( x
+            , Prim (Extern "caml_alloc_dummy_function", [ size; Pc (Int (Int32.succ a)) ])
+            )
+      | _ -> assert false)
+  | Let (_, (Apply _ | Prim (Extern ("%resume" | "%perform"), _))) -> assert false
```

Comment:
We don't really need to since it can only occur in tail position in the standard library, in functions that cannot be inlined.
But it does not cost much and can only make the code more robust by including it as well.